node/node_handler.go

```diff
@@ -189,27 +202,136 @@ func (node *Node) BroadcastCrossLink() {
 		"Construct and Broadcasting new crosslink to beacon chain groupID %s",
 		nodeconfig.NewGroupIDByShardID(shard.BeaconChainShardID),
 	)
-	headers := []*block.Header{}
-	lastLink, err := node.Beaconchain().ReadShardLastCrossLink(curBlock.ShardID())
+
+	headers, err := getCrosslinkHeadersForShards(node.Beaconchain(), node.Blockchain(), curBlock, shardID)
+	if err != nil {
+		utils.Logger().Error().Err(err).Msg("[BroadcastCrossLink] failed to get crosslinks")
+		return
+	}
+
+	if len(headers) == 0 {
+		utils.Logger().Info().Msg("[BroadcastCrossLink] no crosslinks to broadcast")
+		return
+	}
+
+	node.host.SendMessageToGroups(
+		[]nodeconfig.GroupID{nodeconfig.NewGroupIDByShardID(shard.BeaconChainShardID)},
+		p2p.ConstructMessage(
+			proto_node.ConstructCrossLinkMessage(node.Consensus.Blockchain, headers)),
+	)
+}
+
+// BroadcastCrossLinkFromBeaconToShards is called by consensus leader or 1% validators to
+// send last cross link to shard chains.
+func (node *Node) BroadcastCrossLinkFromBeaconToShards() { // leader of 0 shard
+	if !node.IsRunningBeaconChain() {
+		return
+	}
+	if !(node.IsCurrentlyLeader() || rand.Intn(100) == 0) {
```

Comment:
better move this condition on the caller of BroadcastCros.....(), we want the method to be simple on the functionality and not involve too much rules on the caller logic.

Reply:
Actually no, caller doesn't know anything about logic, it's better lo leave it here, otherwise we have to move amount of logic from class that knows everything into caller, that knows nothing.

Reply:
OK, works too.